graphbolt/src/csc_sampling_graph.cc

```diff
@@ -131,16 +131,150 @@ c10::intrusive_ptr<SampledSubgraph> CSCSamplingGraph::InSubgraph(
+               int64_t offset, int64_t num_neighbors) {
+      int64_t max_count = fanout == -1 ? num_neighbors : fanout;
+      int64_t nonzero_count = 0;
+      nonzero_count = torch::nonzero(probs_or_mask.value().slice(
+                                         0, offset, offset + num_neighbors))
+                          .size(0);
+      if (replace) {
+        return torch::tensor({nonzero_count == 0 ? 0 : max_count});
+      } else {
+        return torch::tensor({std::min(nonzero_count, max_count)});
+      }
+    };
+  }
+}
+
+NumPickFn GetNumPickFnByEtype(
+    const std::vector<int64_t>& fanouts, bool replace,
+    const torch::Tensor& type_per_edge,
+    const torch::optional<torch::Tensor>& probs_or_mask) {
+  return [replace, fanouts, probs_or_mask, &type_per_edge](
+             int64_t offset, int64_t num_neighbors) {
+    int64_t etype_begin = offset;
+    int64_t etype_end = offset;
+    const int64_t end = offset + num_neighbors;
+    torch::Tensor count_by_etype = torch::zeros(fanouts.size());
+    AT_DISPATCH_ALL_TYPES(
+        type_per_edge.scalar_type(), "NumPickFnByEtype", ([&] {
+          const scalar_t* type_per_edge_data =
+              type_per_edge.data_ptr<scalar_t>();
+          while (etype_begin < end) {
+            scalar_t etype = type_per_edge_data[etype_begin];
+            TORCH_CHECK(
+                etype >= 0 && etype < (int64_t)fanouts.size(),
+                "Etype values exceed the number of fanouts.");
+            auto etype_end_it = std::upper_bound(
+                type_per_edge_data + etype_begin, type_per_edge_data + end,
+                etype);
+            etype_end = etype_end_it - type_per_edge_data;
+            // Do sampling for one etype.
+            count_by_etype[etype] =
+                GetNumPickFn(fanouts[etype], replace, probs_or_mask)(
+                    etype_begin, etype_end - etype_begin)[0];
+            etype_begin = etype_end;
+          }
+        }));
+    return count_by_etype;
+  };
+}
+
+/**
+ * @brief Get a lambda function which counts the number of the neighbors to be
+ * sampled.
+ *
+ * @param fanouts The number of edges to be sampled for each node with or
+ * without considering edge types.
+ * @param replace Boolean indicating whether the sample is performed with or
+ * without replacement. If True, a value can be selected multiple times.
+ * Otherwise, each value can be selected only once.
+ * @param type_per_edge A tensor representing the type of each edge, if
+ * present.
+ * @param probs_or_mask Optional tensor containing the (unnormalized)
+ * probabilities associated with each neighboring edge of a node in the original
+ * graph. It must be a 1D floating-point tensor with the number of elements
+ * equal to the number of edges in the graph.
+ *
+ * @return A lambda function which takes offset and num_neighbors as params and
+ * returns the number of neighbors-to-be-sampled.
+ */
+NumPickFn ChooseNumPickFn(
```

Comment:
These 3 functions of `num_pick` is a bit of confusing, because all of them looks are performing "get" work.  I would suggest the following organization of code:

1. NumPick
2. NumPickByEtype
3. GetNumPickFn